### ISSUE TITLE
servlet rule creating spans with incorrect http.status_code values

### ISSUE THREAD

**Issue by randallt** [opentracing-contrib/java-specialagent#533]
I am currently using a fork of v1.5.8 of the java special agent. 

I noticed that I was never getting anything other than an http status code of 200 from certain services that use the servlet rule. This does not happen on all my services that use this rule, but it does happen on multiple services that have the same technology stack (Spring Boot 2.0.3 and Jetty 9.4.11)

I have a sample app that just returns a 401 every time, as shown by curl. However, the spans created have a https status code of 200. I instrumented io.opentracing.contrib.specialagent.rule.servlet.FilterAgentIntercept.getStatusCode and found that the ServletResponse passed in is never in the WeakHashMap, so a 200 is always returned. I also instrumented setStatusCode and found that it is never being called.

I compared the current v1.7.0 code of the servlet rule and I didn't see differences that would fix this, but let me know if you feel differently.

**Comment by safris** [comment]
Hi @randallt, is this a new issue, or are you referring to something that you've mentioned before? I don't seem to remember something related to the `servlet` rule always returning 200. If this is indeed an issue, your sample app would be very helpful.

**Comment by randallt** [comment]
This is a new issue that we just discovered today.

Randall Theobald
Sr Principal Software Engineer
Performance Engineering • Product Engineering
Secureworks® | Collectively Smarter. Exponentially Safer.™
rtheobald@secureworks.com
________________________________

**Comment by safris** [comment]
If you could provide your sample app to demonstrate this issue, I would greatly appreciate it!

**Comment by randallt** [comment]
Unfortunately, I shouldn't have called it a sample app, as it contains components that I can't upload due to business restrictions. I will see if we can trim it down. But in the meantime, is there debugging output that would be helpful to diagnose the issue? I am also in the process of syncing my fork with the 1.7.0 version of the agent.

**Comment by safris** [comment]
No problem. And I'll try to tweak the integration tests to simulate non-200 responses. If you turn on `-Dsa.log.level=FINER`, you would see the method intercepts resulting in `ServletContext.addFilter()` calls. Unfortunately, I don't believe the `web-servlet-filter` integration has debug log statements. I'll look through the code to see if I can add some.

**Comment by randallt** [comment]
I merged with the 1.7.0 release and am still seeing the issue. Also, FYI, I saw an increase in the start time, not a decrease like I was hoping. But nothing drastic. I also added the spring boot and spring webmvc plugins to my build to see if deferred static attach would help. It did not, and I also see this in the console:
```
<><><><> Multiple integrations registered for: class org.springframework.web.servlet.DispatcherServlet   pluginManifest: spring:webmvc
<><><><> Multiple integrations registered for: class org.springframework.web.servlet.DispatcherServlet   pluginManifest: spring:webmvc
<><><><> Multiple integrations registered for: class org.springframework.web.servlet.FrameworkServlet   pluginManifest: spring:webmvc
<><><><> Multiple integrations registered for: class org.springframework.web.servlet.FrameworkServlet   pluginManifest: spring:webmvc
```
I added the `pluginManifest: ...` part to see which plugin was being reported. I guess this is a separate issue that you might want to look into, as the behavior I'm reporting in this issue seem the same either way.

We should have a cleaned sample application for you today, and I will try to get the FINER log.

**Comment by randallt** [comment]
FINER log attached of startup and a single request that resulted in a 401 (but the span says 200) http status code.
[sa.log.gz](https://github.com/opentracing-contrib/java-specialagent/files/4451196/sa.log.gz)

**Comment by randallt** [comment]
Sample application that I confirm reproduces the issue attached:

[sa.issue-533.sample.zip](https://github.com/opentracing-contrib/java-specialagent/files/4451576/sa.issue-533.sample.zip)

It is large because we included a built JAR. To start the server, just run
```
java -jar ct-helloworld.jar --spring.jersey.application-path=demo
```
then you can get different http status codes by using curl like:
```
curl -v localhost:8080/demo/simerror/503
```

**Comment by randallt** [comment]
But man, we've got to do something about the startup time using the agent. Even with my trimmed down version, server startup goes from 4 sec to 24 sec (and our main webapp goes from about 30 sec to 120 or so). Stop time is also affected (I'm guessing the tmp file delete shutdown hook). Is there any possibility to not write so many temporary class files? I might be able to look into it if there is a chance.

**Comment by safris** [comment]
Hi @randallt, I am surprised to hear that the startup time has regressed. I will look at your sample app to try to understand what is happening.

**Comment by randallt** [comment]
The incorrect HTTP status is definitely higher priority than start time. I think the slower startup could be due to my moving my Wavefront tracer plugin to be in the ‘iso’ plugin directory.

Randall Theobald
Sr Principal Software Engineer
Performance Engineering • Product Engineering
Secureworks® | Collectively Smarter. Exponentially Safer.™
rtheobald@secureworks.com
________________________________

**Comment by safris** [comment]
Hi @randallt, thank you for the sample app. I was able to use it to figure out what was happening, and have fixed the issue. In short, this is all due to the fact that `HttpServletResponse` in Servlet API _before_ v3+ did not have `getStatus()` method (what the hell was Sun thinking for leaving it out?!). I debugged Glassfish to see how it sets the status code, and apparently it uses a [deprecated method](https://github.com/jersey/jersey/blob/master/containers/jersey-servlet-core/src/main/java/org/glassfish/jersey/servlet/internal/ResponseWriter.java#L171) that was not instrumented in SpecialAgent. In addition to this, the behavior of SpecialAgent was to return status code 200 if the code was missing, which is a mistake. I've updated the code to be more responsive to the version of the Servlet API, as well as to instrument the deprecated `setStatus(int,String)` method. I've also added `FINER` level logging to show what status code is submitted to the span. Please rebase and rebuild the `servlet` rule in your codebase, and let me know if this fixes your issue.

**Comment by randallt** [comment]
This fixes the issue. How often do you expect the reflection code you added to be called?

**Comment by safris** [comment]
> How often do you expect the reflection code you added to be called?
Never, but I guess I'm wrong.

I used [this StackOverflow post](https://stackoverflow.com/questions/1302072/how-can-i-get-the-http-status-code-out-of-a-servletresponse-in-a-servletfilter) to guide me as per the methods that should be instrumented. As far as I can see, [these methods](https://github.com/opentracing-contrib/java-specialagent/blob/master/rule/servlet/src/main/java/io/opentracing/contrib/specialagent/rule/servlet/FilterAgentRule.java#L47-L52) should cover all the ways a status can be set in an `HttpServletResponse`.

**Comment by randallt** [comment]
Maybe it does need to default to 200? So far I only see it using reflection for 200s, but my testing hasn't been comprehensive.

**Comment by randallt** [comment]
Yeah, I got the runtime class of the offending service execution and it sets the status code to 200 by default:
https://github.com/eclipse/jetty.project/blob/jetty-9.4.22.v20191022/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java#L91

**Comment by randallt** [comment]
It seems to work, but I have discovered that I'm completely missing some requests with this service. This particular service is a Jetty server with an embedded WAR. For things in the WAR's context, I seem to get them (this is obviously the most important). But for 404s resulting from requests outside of the context, it looks like the instrumentation isn't picking it up at all. Would that possibly require custom Jetty rules?

**Comment by randallt** [comment]
Looks like we're not intercepting requests that don't match a jetty context handler because Jetty sets the error and avoids the filter chain altogether. I suppose that is fine. Consider my problem fixed, but you might want to consider a solution to the "too much reflection" issue I noticed (maybe open a follow-on before closing this).

**Comment by safris** [comment]
Thanks for the thorough follow-up with this @randallt.

> Maybe it does need to default to 200?

Now that you mention this, I do seem to remember why I added that in the first place -- because if you don't call `HttpServletResponse.setStatus(...)`, then it defaults to `200`. I think this actually does mean that defaulting to 200 is correct, given that all ways of setting a status in a `HttpServletResponse` are properly captured by the rule, which I believe they are (now). I will update the code to fallback to 200 if no explicit status was set.

I've followed your suggestion, and have created a new issue to track "too much reflection": #547.
